File: problems/climbing_stairs/ClimbingStairs.java
package problems.climbing_stairs;

import java.util.Map;
import java.util.HashMap;

public class ClimbingStairs {
    private Map<String, Integer> memo = new HashMap<>();

    public int climbStairs(int n) {
        if (n == 0)
            return 0;
        if (n == 1)
            return 1;
        if (n == 2)
            return 2;
        String key = String.valueOf(n) + "++";
        if (memo.containsKey(key))
            return memo.get(key);

        int answer = climbStairs(n - 1) + climbStairs(n - 2);
        memo.put(key, answer);
        return memo.get(key);
    }
}

File: problems/climbing_stairs/climbing_stairs.py
memo = {}


class Solution:
    def climbStairs(self, n: int) -> int:
        if n == 0:
            return 0
        if n == 1:
            return 1
        if n == 2:
            return 2
        key = f"{n}++"
        if key in memo:
            return memo[key]

        answer = self.climbStairs(n - 1) + self.climbStairs(n - 2)
        memo[key] = answer
        return memo[key]
